kuka_rsi_hw_interface/src/robot_control_node.cpp

```diff
@@ -22,76 +22,134 @@
+                                   kroshu_ros2_core::ParameterSetAccessRights{true, true, false, false},
+                                   [this](const std::string &rsi_ip_address)
+                                   { return this->onRSIIPAddressChange(rsi_ip_address); });
+
+    registerParameter<int>("rsi_port_", 59152,
+                           kroshu_ros2_core::ParameterSetAccessRights{true, true, false, false},
+                           [this](int rsi_port)
+                           { return this->onRSIPortAddressChange(rsi_port); });
+
+    registerParameter<int>("n_dof_", 6,
+                           kroshu_ros2_core::ParameterSetAccessRights{true, true, false, false},
+                           [this](int n_dof)
+                           { return this->onNDOFChange(n_dof); });
+
+    kuka_rsi_hw_interface_ = std::make_unique<KukaHardwareInterface>(rsi_ip_address_, rsi_port_, n_dof_);
+    return SUCCESS;
+  }
+  CallbackReturn RobotControlNode::on_activate(const rclcpp_lifecycle::State &)
+  {
+    kuka_rsi_hw_interface_->start(joint_state_msg_.position, joint_command_msg_->position, initial_joint_pos_);
 
-rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn RobotControlNode::
-on_activate(
-  const rclcpp_lifecycle::State &)
-{
-  kuka_rsi_hw_interface_->start();
+    joint_state_publisher_->on_activate();
+    joint_state_publisher_->publish(joint_state_msg_);
 
-  control_thread_ = std::thread(&RobotControlNode::Controlloop, this);
-  return SUCCESS;
-}
+    control_thread_ = std::thread(&RobotControlNode::ControlLoop, this);
+    return SUCCESS;
+  }
 
-rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn RobotControlNode::
-on_deactivate(
-  const rclcpp_lifecycle::State &)
-{
-  kuka_rsi_hw_interface_->stop();
-  if (control_thread_.joinable()) {control_thread_.join();}
+  CallbackReturn RobotControlNode::on_deactivate(const rclcpp_lifecycle::State &)
+  {
+    kuka_rsi_hw_interface_->stop(joint_pos_correction_deg_);
+    joint_state_publisher_->on_deactivate();
+    if (control_thread_.joinable())
+    {
+      control_thread_.join();
+    }
 
-  return SUCCESS;
-}
+    return SUCCESS;
+  }
 
-rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn RobotControlNode::
-on_cleanup(
-  const rclcpp_lifecycle::State &)
-{
-  kuka_rsi_hw_interface_->cleanup();
-  return SUCCESS;
-}
+  CallbackReturn RobotControlNode::on_cleanup(const rclcpp_lifecycle::State &)
+  {
+    kuka_rsi_hw_interface_ = nullptr;
+    return SUCCESS;
+  }
 
-void RobotControlNode::Controlloop()
-{
-  while (kuka_rsi_hw_interface_->isActive()) {
-    if (!kuka_rsi_hw_interface_->read()) {
-      RCLCPP_ERROR(get_logger(), "Failed to read state from robot. Shutting down!");
-      rclcpp::shutdown();
-      return;
+  void RobotControlNode::ControlLoop()
+  {
+    while (kuka_rsi_hw_interface_->isActive())
+    {
+      std::unique_lock<std::mutex> lock(m_);
```

Comment:
I would not put the mutex lock before the read() function, as it is a blocking call and thus commandReceivedCallback may not be allowed to run at all